handlers/aws/handler.py

```diff
@@ -81,22 +81,41 @@ def lambda_handler(lambda_event: dict[str, Any], lambda_context: context_.Contex
     if trigger_type == "replay-sqs":
         replay_queue_arn = lambda_event["Records"][0]["eventSourceARN"]
         replay_handler = ReplayedEventReplayHandler(replay_queue_arn=replay_queue_arn)
+        shipper_cache: dict[str, ProtocolShipper] = {}
         for replay_record in lambda_event["Records"]:
             event = json_parser(replay_record["body"])
-            _handle_replay_event(
-                config=config,
-                output_type=event["output_type"],
-                output_args=event["output_args"],
-                event_input_id=event["event_input_id"],
-                event_payload=event["event_payload"],
-                replay_handler=replay_handler,
-                receipt_handle=replay_record["receiptHandle"],
+            input_id = event["event_input_id"]
+            if input_id not in shipper_cache:
+                shipper = get_shipper_for_replay_event(
+                    config=config,
+                    output_type=event["output_type"],
+                    output_args=event["output_args"],
+                    event_input_id=event["event_input_id"],
+                    replay_handler=replay_handler,
+                )
+
+                if shipper is None:
+                    shared_logger.warn(
+                        "no shipper for output in replay queue",
+                        extra={"output_type": event["output_type"], "event_input_id": event["event_input_id"]},
+                    )
+                    continue
+
+                shipper_cache[input_id] = shipper
+            else:
+                shipper = shipper_cache[input_id]
```

Comment:
under what circumstances would different events require different shippers?

Reply:
when the replay messages come from different inputs with different outputs

Reply:
got it - if we have a single lambda deployment configured, do we still end up constructing multiple shippers here, one per batch of input events, even though the configuration of the shipper is identical? 

i don't see any issues, just trying to understand if there's an optimization we could make (for example change the caching logic to be based on the shipper config, vs the input id).

Reply:
> i don't see any issues, just trying to understand if there's an optimization we could make (for example change the caching logic to be based on the shipper config, vs the input id).

yes, this is an optimisation that could be made, but most probably [here](https://github.com/elastic/elastic-serverless-forwarder/blob/main/handlers/aws/handler.py#L372-L382), since in the replay queue the same shipper (as es cluster with credentials) can have different [es_datastream_name](https://github.com/elastic/elastic-serverless-forwarder/pull/157/files#diff-a50c2cea8bcb0bea90c7963fe6a2ecb9a44fe4a04cd8738ce7d9f626d2109ea6R54), according to the original input

Reply:
> yes, this is an optimisation that could be made, but most probably [here](https://github.com/elastic/elastic-serverless-forwarder/blob/main/handlers/aws/handler.py#L372-L382)

even there the same related to `es_datastream_name` is true.
it would be relevant refactoring to try to optimise the shipper cache removing `es_datastream_name` from the picture, not sure if it is worth